samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -43,53 +78,259 @@
+  }
+
+  /**
+   *  Creates SystemConsumer objects for store restoration, creating one consumer per system.
+   */
+  private Map<String, SystemConsumer> createStoreConsumers(Map<String, SystemStream> changelogSystemStreams,
+      Map<String, SystemFactory> systemFactories, Config config) {
+    // Determine the set of systems being used across all stores
+    Set<String> storeSystems =
+        changelogSystemStreams.values().stream().map(SystemStream::getSystem).collect(Collectors.toSet());
+
+    // Create one consumer for each system in use
+    Map<String, SystemConsumer> storeSystemConsumers = new HashMap<>();
+
+    for (String storeSystemName : storeSystems) {
+      SystemFactory systemFactory = systemFactories.get(storeSystemName);
+      if (systemFactory == null) {
+        throw new SamzaException("Changelog system " + storeSystemName + " does not exist in config");
+      }
+      storeSystemConsumers.put(storeSystemName,
+          systemFactory.getConsumer(storeSystemName, config, samzaContainerMetrics.registry()));
+    }
+
+    // Populate the map of storeName to its relevant systemConsumer
+    return changelogSystemStreams.entrySet()
+        .stream()
+        .collect(Collectors.toMap(Map.Entry::getKey, x -> storeSystemConsumers.get(x.getValue().getSystem())));
+  }
+
+  /**
+   * Create taskStores with the given store mode.
+   * In case a store already exists in taskStores and it is a persistent store, the existing store is overwritten with a new one.
+   * Existing non-persistent stores in taskStores are left as-is.
+   *
+   */
+  private void createTaskStores(ContainerModel containerModel, JobContext jobContext, ContainerContext containerContext,
+      Map<String, StorageEngineFactory<Object, Object>> storageEngineFactories,
+      Map<String, SystemStream> changelogSystemStreams, Map<String, Serde<Object>> serdes,
+      Map<TaskName, TaskInstanceMetrics> taskInstanceMetrics,
+      Map<TaskName, TaskInstanceCollector> taskInstanceCollectors, StorageEngineFactory.StoreMode storeMode) {
+
+    // iterate over each task and each storeName
+    for (Map.Entry<TaskName, TaskModel> task : containerModel.getTasks().entrySet()) {
+      TaskName taskName = task.getKey();
+      TaskModel taskModel = task.getValue();
+
+      for (String storeName : storageEngineFactories.keySet()) {
+
+        if (this.taskStores.get(taskName) == null) {
+          this.taskStores.put(taskName, new HashMap<>());
+        }
+
+        // if this store has been already created in the taskStores, then re-create and overwrite it only if it is a persistentStore
+        if (this.taskStores.get(taskName).containsKey(storeName) && this.taskStores.get(taskName)
+            .get(storeName)
+            .getStoreProperties()
+            .isPersistedToDisk()) {
+
+          StorageEngine storageEngine =
+              createStore(storeName, taskName, taskModel, jobContext, containerContext, storageEngineFactories,
```

Comment:
Currently,`createTaskStores` is currently invoked twice - once during construction and another during `start`.. 

Instead, would prefer to have an explicit method - `recreatePersistentStoresInRWMode()` instead of having complexity hidden in `createStores`.

Reply:
done